Can kill OpenShift process with go panic by invalid json file

Env
---
~~~
# openshift version
openshift v1.0.5-27-g2cc96dc
kubernetes v1.1.0-alpha.0-1605-g44c91b1

# go version
go version go1.4.2 linux/amd64
~~~

**NOTE:** I couldn't reproduce this issue with OSE v3.0.1.0 due to another error.

Issue:
---
- We can kill OpenShift process by sending request with invalid json file.
- Please see following steps.

Reproduce steps
---

step-1. Start openshift origin with standalone mode, and create route and docker-registry. (Followed with [here](https://github.com/nak3/openshift-local-setup))

step-2. Send invalid json file by my [origin-killer.json](https://gist.githubusercontent.com/nak3/23de3850fd3abf5ed038/raw/eba398a3e778459ca21ff8f60371e708566f7950/origin-killer.json
)
~~~
[root@localhost ~]# oc create -f https://gist.githubusercontent.com/nak3/23de3850fd3abf5ed038/raw/eba398a3e778459ca21ff8f60371e708566f7950/origin-killer.json
imagestream "sti-python" created
Unable to connect to the server: EOF
The connection to the server 192.168.122.27:8443 was refused - did you specify the right host or port?
The connection to the server 192.168.122.27:8443 was refused - did you specify the right host or port?
~~~

step-3. You can see go panic and stop the OpenShift process

The log is here: https://kenjiro.fedorapeople.org/misc/logs/origin.log (Too long, please check from the bottom.)

```
I0826 16:21:40.165796    2410 image_change_controller.go:47] Build image change controller detected ImageStream change 172.30.157.204:5000/default/sti-python
I0826 16:21:40.166325    2410 controller.go:38] Detecting changed images for DeploymentConfig default/docker-registry:1
I0826 16:21:40.166483    2410 controller.go:38] Detecting changed images for DeploymentConfig default/router:1
I0826 16:21:40.166497    2410 controller.go:86] Updated all DeploymentConfigs for trigger on ImageStream default/sti-python
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x66bbb8]

goroutine 6548 [running]:
github.com/openshift/origin/pkg/build/util.GetImageStreamForStrategy(0xc20a2fadda, 0x6, 0x0, 0x0, 0xc20e0d3720, 0x401ed10)
	/tmp/origin/_output/local/go/src/github.com/openshift/origin/pkg/build/util/util.go:37 +0x128
github.com/openshift/origin/pkg/build/api/validation.ValidateBuildConfig(0xc20bf47200, 0x0, 0x0, 0x0)
	/tmp/origin/_output/local/go/src/github.com/openshift/origin/pkg/build/api/validation/validation.go:67 +0x5e8
github.com/openshift/origin/pkg/build/registry/buildconfig.strategy.Validate(0x7f6b977fd3c8, 0xc2080d4af0, 0x7f6b977fb558, 0x401ed10, 0x7f6b9781cd30, 0xc20d53a090, 0x7f6b977fe438, 0xc20bf47200, 0x0, 0x0, ...)
	/tmp/origin/_output/local/go/src/github.com/openshift/origin/pkg/build/registry/buildconfig/strategy.go:51 +0x67
github.com/openshift/origin/pkg/build/registry/buildconfig.(*strategy).Validate(0xc208f535c0, 0x7f6b9781cd30, 0xc20d53a090, 0x7f6b977fe438, 0xc20bf47200, 0x0, 0x0, 0x0)
	<autogenerated>:32 +0xef
k8s.io/kubernetes/pkg/api/rest.BeforeCreate(0x7f6b97672698, 0xc208f535c0, 0x7f6b9781cd30, 0xc20d53a090, 0x7f6b977fe438, 0xc20bf47200, 0x0, 0x0)
	/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/rest/create.go:66 +0x367
k8s.io/kubernetes/pkg/registry/generic/etcd.(*Etcd).Create(0xc208e45a20, 0x7f6b9781cd30, 0xc20d53a090, 0x7f6b977fe438, 0xc20bf47200, 0x0, 0x0, 0x0, 0x0)
	/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go:205 +0x193
k8s.io/kubernetes/pkg/apiserver.(*namedCreaterAdapter).Create(0xc2093a3970, 0x7f6b9781cd30, 0xc20d53a090, 0x0, 0x0, 0x7f6b977fe438, 0xc20bf47200, 0x0, 0x0, 0x0, ...)
	/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:357 +0xa2
k8s.io/kubernetes/pkg/apiserver.func·025(0x0, 0x0, 0x0, 0x0)
	/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:322 +0xaa
k8s.io/kubernetes/pkg/apiserver.func·033()
	/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:622 +0x38
created by k8s.io/kubernetes/pkg/apiserver.finishRequest
	/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:627 +0x1e4
```

--- a/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go
+++ b/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go
@@ -25,6 +25,7 @@ import (
 	"net"
 	"net/http"
 	"path"
+	rt "runtime"
 	"strconv"
 	"strings"
 	"time"
@@ -155,6 +156,27 @@ func InstallLogsSupport(mux Mux) {
 	// TODO: use restful: ws.Route(ws.GET("/logs/{logpath:*}").To(fileHandler))
 	// See github.com/emicklei/go-restful/blob/master/examples/restful-serve-static.go
 	mux.Handle("/logs/", http.StripPrefix("/logs/", http.FileServer(http.Dir("/var/log/"))))
+}
+
+func InstallRecoverHandler(container *restful.Container) {
+	container.RecoverHandler(logStackOnRecover)
+}
+
+//TODO: Unify with RecoverPanics?
+func logStackOnRecover(panicReason interface{}, httpWriter http.ResponseWriter) {
+	var buffer bytes.Buffer
+	buffer.WriteString(fmt.Sprintf("recover from panic situation: - %v\r\n", panicReason))
+	for i := 2; ; i += 1 {
+		_, file, line, ok := rt.Caller(i)
+		if !ok {
+			break
+		}
+		buffer.WriteString(fmt.Sprintf("    %s:%d\r\n", file, line))
+	}
+	glog.Errorln(buffer.String())
+
+	// TODO: make status unversioned or plumb enough of the request to deduce the requested API version
+	errorJSON(apierrors.NewGenericServerResponse(http.StatusInternalServerError, "", "", "", "", 0, false), latest.Codec, httpWriter)
 }
 
 func InstallServiceErrorHandler(container *restful.Container, requestResolver *APIRequestInfoResolver, apiVersions []string) {

--- a/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go
+++ b/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go
@@ -29,6 +29,7 @@ import (
 	"k8s.io/kubernetes/pkg/api/rest"
 	"k8s.io/kubernetes/pkg/fields"
 	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util"
 	"k8s.io/kubernetes/pkg/util/strategicpatch"
 
 	"github.com/emicklei/go-restful"
@@ -618,7 +619,14 @@ func finishRequest(timeout time.Duration, fn resultFunc) (result runtime.Object,
 	// when the select statement reads something other than the one the goroutine sends on.
 	ch := make(chan runtime.Object, 1)
 	errCh := make(chan error, 1)
+	panicCh := make(chan interface{}, 1)
 	go func() {
+		// panics don't cross goroutine boundaries, so we have to handle ourselves
+		defer util.HandleCrash(func(panicReason interface{}) {
+			// Propagate to parent goroutine
+			panicCh <- panicReason
+		})
+
 		if result, err := fn(); err != nil {
 			errCh <- err
 		} else {
@@ -634,6 +642,8 @@ func finishRequest(timeout time.Duration, fn resultFunc) (result runtime.Object,
 		return result, nil
 	case err = <-errCh:
 		return nil, err
+	case p := <-panicCh:
+		panic(p)
 	case <-time.After(timeout):
 		return nil, errors.NewTimeoutError("request did not complete within allowed duration", 0)
 	}

--- a/Godeps/_workspace/src/k8s.io/kubernetes/pkg/master/master.go
+++ b/Godeps/_workspace/src/k8s.io/kubernetes/pkg/master/master.go
@@ -17,7 +17,6 @@ limitations under the License.
 package master
 
 import (
-	"bytes"
 	"fmt"
 	"io/ioutil"
 	"math/rand"
@@ -26,7 +25,6 @@ import (
 	"net/http/pprof"
 	"net/url"
 	"os"
-	rt "runtime"
 	"strconv"
 	"strings"
 	"sync"
@@ -412,22 +410,8 @@ func (m *Master) HandleFuncWithAuth(pattern string, handler func(http.ResponseWr
 func NewHandlerContainer(mux *http.ServeMux) *restful.Container {
 	container := restful.NewContainer()
 	container.ServeMux = mux
-	container.RecoverHandler(logStackOnRecover)
+	apiserver.InstallRecoverHandler(container)
 	return container
-}
-
-//TODO: Unify with RecoverPanics?
-func logStackOnRecover(panicReason interface{}, httpWriter http.ResponseWriter) {
-	var buffer bytes.Buffer
-	buffer.WriteString(fmt.Sprintf("recover from panic situation: - %v\r\n", panicReason))
-	for i := 2; ; i += 1 {
-		_, file, line, ok := rt.Caller(i)
-		if !ok {
-			break
-		}
-		buffer.WriteString(fmt.Sprintf("    %s:%d\r\n", file, line))
-	}
-	glog.Errorln(buffer.String())
 }
 
 // init initializes master.

--- a/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/util.go
+++ b/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/util.go
@@ -44,12 +44,16 @@ var ReallyCrash bool
 var PanicHandlers = []func(interface{}){logPanic}
 
 // HandleCrash simply catches a crash and logs an error. Meant to be called via defer.
-func HandleCrash() {
+// Additional context-specific handlers can be provided, and will be called in case of panic
+func HandleCrash(additionalHandlers ...func(interface{})) {
 	if ReallyCrash {
 		return
 	}
 	if r := recover(); r != nil {
 		for _, fn := range PanicHandlers {
+			fn(r)
+		}
+		for _, fn := range additionalHandlers {
 			fn(r)
 		}
 	}

--- a/pkg/build/util/util.go
+++ b/pkg/build/util/util.go
@@ -34,10 +34,19 @@ func GetBuildName(pod *kapi.Pod) string {
 func GetImageStreamForStrategy(strategy buildapi.BuildStrategy) *kapi.ObjectReference {
 	switch strategy.Type {
 	case buildapi.SourceBuildStrategyType:
+		if strategy.SourceStrategy == nil {
+			return nil
+		}
 		return &strategy.SourceStrategy.From
 	case buildapi.DockerBuildStrategyType:
+		if strategy.DockerStrategy == nil {
+			return nil
+		}
 		return strategy.DockerStrategy.From
 	case buildapi.CustomBuildStrategyType:
+		if strategy.CustomStrategy == nil {
+			return nil
+		}
 		return &strategy.CustomStrategy.From
 	default:
 		return nil

